cpp/src/table/row_operators.cu

```diff
@@ -402,126 +483,119 @@ namespace {
  *
- * @param lhs The input lhs column to transform
- * @param rhs The input rhs column to transform (if available)
+ * @param input The input column to transform
  * @param column_null_order The flag indicating how nulls compare to non-null values
  * @param stream CUDA stream used for device memory operations and kernel launches
- * @return A tuple consisting of new column_view representing the transformed input, along with
- *         their ranks column(s) (of `size_type` type) and possibly new list offsets generated
+ * @param mr Device memory resource used to allocate the returned column(s)
+ * @return A pair consisting of new column_view representing the transformed input, along with
+ *         an array containing its rank column(s) (of `size_type` type) and possibly new list
+ *         offsets generated during the transformation process
+ */
+std::pair<column_view, std::vector<std::unique_ptr<column>>> transform_lists_of_structs(
+  column_view const& input,
+  null_order column_null_order,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  std::vector<std::unique_ptr<column>> out_cols;
+
+  if (input.type().id() == type_id::LIST) {
+    auto const child = cudf::lists_column_view{input}.get_sliced_child(stream);
+
+    // Found a lists-of-structs column.
+    if (child.type().id() == type_id::STRUCT) {
+      out_cols.emplace_back(compute_ranks(child, column_null_order, stream, mr));
+      auto transformed = replace_child(input, out_cols.back()->view(), out_cols, stream, mr);
+      return {std::move(transformed), std::move(out_cols)};
+    }
+    // Found a lists-of-lists column.
+    else if (child.type().id() == type_id::LIST) {
+      // Recursively call transformation on the child column.
+      auto [new_child, out_cols_child] =
+        transform_lists_of_structs(child, column_null_order, stream, mr);
+
+      // Only transform the current column if its child has been transformed.
+      if (out_cols_child.size() > 0) {
+        out_cols.insert(out_cols.end(),
+                        std::make_move_iterator(out_cols_child.begin()),
+                        std::make_move_iterator(out_cols_child.end()));
+        auto transformed = replace_child(input, new_child, out_cols, stream, mr);
+        return {std::move(transformed), std::move(out_cols)};
+      }
+      // else: child was not transformed so input is also not transformed.
+    }
+    // else: child is not STRUCT or LIST: no transformation.
+  }
+  // else: lhs.type().id() != type_id::LIST.
+  // In such situations, lhs.type().id() can still be type_id::STRUCT. However, any
+  // structs-of-lists should be decomposed into empty struct type `Struct<>` before being
+  // processed by this function so we do nothing here.
+
+  // Passthrough: nothing changed.
+  return {input, std::move(out_cols)};
+}
+
+/**
+ * @copydoc transform_lists_of_structs(column_view const&, null_order, rmm::cuda_stream_view)
```

Comment:
This signature is no longer correct since you've added an `mr` parameter above. However, it also seems like it would be incorrect because the other function has an `input` column view whereas this function has `lhs` and `rhs`, so using copydoc will end up showing 3 input columns.